### PR TITLE
Response validation level fail

### PR DESCRIPTION
**Pull request by ryshep111**
Added new response validation level 'fail' that changes the actual http response code to 522 and separates out the validation errors and response body. May break client code, so this should only be used when developing/testing an API in stand alone.

### CODE REVIEW

File: handlers/_swagger.js [hunk 1]
```diff
@@ -259,48 +260,46 @@ function registerRoute(app, auth, additionalMiddleware, method, path, data, allo
...
                             statusCode: res.statusCode,
                             body: invalidBody || body
                         };
-                        logger[responseModelValidationLevel]('Response validation error:', JSON.stringify(validationErrors, null, 2));
+                        logger[responseModelValidationLevel === 'warn' ? 'warn' : 'error']('Response validation error:', JSON.stringify(validationErrors, null, 2));
                     }
+
+                    //after this initial call (sometimes `send` will call itself again), we don't need to get the response for validation anymore
+                    res.send = responseSender;
+                    responseSender.call(res, isJson ? JSON.stringify(body): body);
```
Comment (seanpk) [inline handlers/_swagger.js]: there should be no need to stringify an object to send it
was the thinking just to put back into `send` exactly the same format of thing we intercepted?

that's not a bad thought, but it is also making me think that we should take more care at the beginning, including:
- handling non-object and non-string parameters
- catching errors thrown by that original `JSON.parse`

File: handlers/_swagger.js [hunk 2]
```diff
@@ -259,48 +260,46 @@ function registerRoute(app, auth, additionalMiddleware, method, path, data, allo
                     var isJson = typeof body === 'string'; //body can come in as JSON or object, we want object
                     body = isJson ? JSON.parse(body) : body;
```
Comment (seanpk) [inline handlers/_swagger.js]: there is more safety we should build in around this ... see the comment on the send below

File: handlers/_swagger.js [hunk 3]
```diff
@@ -19,7 +19,8 @@ var polymorphicValidation;
 
 exports.init = function (app, auth, config, logger, serviceLoader, swagger, callback) {
     var cfg = config.get('swagger');
-    responseModelValidationLevel = /error|warn/.test(cfg.validateResponseModels) ? cfg.validateResponseModels : 0;
+    
+    responseModelValidationLevel = /error|warn|fail/.test(cfg.validateResponseModels) ? cfg.validateResponseModels : 0;
```
Comment (seanpk) [inline handlers/_swagger.js]: calculated swagger config could be done once in the swagger service and then exported for use here (and potentially elsewhere)
... see the comment in the swagger service

File: services/swagger.js [hunk 4]
```diff
@@ -21,7 +21,7 @@ var httpMethods = ['get', 'put', 'post', 'delete', 'options', 'head', 'patch'];
 
 exports.init = function(logger, config, callback) {
     var cfg = config.get('swagger');
-    var responseModelValidationLevel = /error|warn/.test(cfg.validateResponseModels) ? cfg.validateResponseModels : 0;
+    var responseModelValidationLevel = /error|warn|fail/.test(cfg.validateResponseModels) ? cfg.validateResponseModels : 0;
```
Comment (seanpk) [inline services/swagger.js]: do something to make this calculated value available as a property of the service
(probably want to calculate the polymorphic validation value here too)

File: test/integration/launchUtil.js [hunk 5]
```diff
@@ -19,26 +19,48 @@ exports.launch = function (fixtureName, opts, done) {
...
-        },
-        function (err, stdout, stderr) {
-            if (err) {
-                console.warn(err, stderr);
-            }
-            output += stdout + stderr;
-        });
+    var bosPath = path.resolve(__dirname, opts.exec);
+    if (process.platform === 'win32') {
+        lastLaunch = child_process.exec('node ' + bosPath,
```
Comment (seanpk) [inline test/integration/launchUtil.js]: does `exec` work on *nix too?
I'm not sure we need this special case for 'win32', I think it will work on both

Comment (ryshep111) [inline test/integration/launchUtil.js]: good point, done

File: services/swagger.js (hunk 4 above)
Comment (ryshep111) [inline services/swagger.js]: done for responseModelValidation, polymorphicValidation and validHttpMethods

File: handlers/_swagger.js (hunk 1 above)
Comment (ryshep111) [inline handlers/_swagger.js]: added try catch, made comments more clear

File: handlers/_swagger.js [hunk 6]
```diff
@@ -259,48 +260,46 @@ function registerRoute(app, auth, additionalMiddleware, method, path, data, allo
...
-                    
-                    if (validationErrors) { // for both errors and warnings ...
                         validationErrors.invalidResponse = {
                             method: req.method,
                             path: req.path,
                             statusCode: res.statusCode,
                             body: invalidBody || body
                         };
-                        logger[responseModelValidationLevel]('Response validation error:', JSON.stringify(validationErrors, null, 2));
+                        logger[responseModelValidationLevel === 'warn' ? 'warn' : 'error']('Response validation error:', JSON.stringify(validationErrors, null, 2));
```
Comment (ryshep111) [inline handlers/_swagger.js]: I broke this into two lines of text for readability, but left it as one line of code

File: handlers/_swagger.js [hunk 7]
```diff
@@ -259,48 +260,46 @@ function registerRoute(app, auth, additionalMiddleware, method, path, data, allo
...
-                        //we'll create a response object (or array entry) if there isn't one (we will break some client code)
-                        validationErrors = validateResponse();
-                        if (validationErrors) {
+                    validationErrors = validateResponseModels(res, body, data, swaggerDoc, logger);
+                    if (validationErrors) {
+                        if (responseModelValidationLevel === 'error' || responseModelValidationLevel === 'fail') {
+                            //we're going to check the model and send any valdiation errors back to the caller in the reponse
+                            //either in the `_response_validation_errors` property of the response, or of that property of the first and last array entries
+                            //we'll create a response object (or array entry) if there isn't one (we will break some client code)
                             invalidBody = _.cloneDeep(body);
```
Comment (ryshep111) [inline handlers/_swagger.js]: done

File: handlers/_swagger.js [hunk 8]
```diff
@@ -259,48 +260,46 @@ function registerRoute(app, auth, additionalMiddleware, method, path, data, allo
...
-                    if (responseModelValidationLevel === 'error') {
-                        //we're going to check the model and send any valdiation errors back to the caller in the reponse
-                        //either in the `_response_validation_errors` property of the response, or of that property of the first and last array entries
-                        //we'll create a response object (or array entry) if there isn't one (we will break some client code)
-                        validationErrors = validateResponse();
-                        if (validationErrors) {
+                    validationErrors = validateResponseModels(res, body, data, swaggerDoc, logger);
+                    if (validationErrors) {
+                        if (responseModelValidationLevel === 'error' || responseModelValidationLevel === 'fail') {
+                            //we're going to check the model and send any valdiation errors back to the caller in the reponse
```
Comment (ryshep111) [inline handlers/_swagger.js]: made comments for each level

File: handlers/_swagger.js [hunk 9]
```diff
@@ -259,48 +260,46 @@ function registerRoute(app, auth, additionalMiddleware, method, path, data, allo
...
-                    var validateResponse = validateResponseModels.bind({},res, body, data, logger, swaggerDoc);
-                    if (responseModelValidationLevel === 'error') {
-                        //we're going to check the model and send any valdiation errors back to the caller in the reponse
-                        //either in the `_response_validation_errors` property of the response, or of that property of the first and last array entries
-                        //we'll create a response object (or array entry) if there isn't one (we will break some client code)
-                        validationErrors = validateResponse();
-                        if (validationErrors) {
+                    validationErrors = validateResponseModels(res, body, data, swaggerDoc, logger);
+                    if (validationErrors) {
+                        if (responseModelValidationLevel === 'error' || responseModelValidationLevel === 'fail') {
```
Comment (ryshep111) [inline handlers/_swagger.js]: done

File: handlers/_swagger.js (hunk 2 above)
Comment (ryshep111) [inline handlers/_swagger.js]: added simple try catch

File: handlers/_swagger.js (hunk 3 above)
Comment (ryshep111) [inline handlers/_swagger.js]: done

**Pull request merged**